File: typescript/src/services/submission-service.ts
import type pg from "pg";
import type { Submission } from "../data/models.js";
import { CampaignState } from "../data/enums.js";
import { SubmissionState } from "../data/enums.js";
import { CampaignRepository } from "../repositories/campaign-repository.js";
import { CreatorRepository } from "../repositories/creator-repository.js";
import { SubmissionRepository } from "../repositories/submission-repository.js";
import { withTransaction } from "../db/pool.js";

export class SubmissionService {
  constructor(private pool: pg.Pool) {}

  async getSubmissionsByCampaign(campaignId: number): Promise<Submission[]> {
    const repo = new SubmissionRepository(this.pool);
    return repo.findByCampaignId(campaignId);
  }

  async getSubmission(id: number): Promise<Submission | null> {
    const repo = new SubmissionRepository(this.pool);
    return repo.findById(id);
  }

  async submitContent(
    campaignId: number,
    creatorId: string,
    contentUrl: string,
  ): Promise<Submission> {
    return withTransaction(this.pool, async (client) => {
      const campaignRepo = new CampaignRepository(client);
      const campaign = await campaignRepo.findById(campaignId);
      if (!campaign) {
        throw new Error(`Campaign ${campaignId} not found`);
      }
      if (campaign.state !== CampaignState.ACTIVE) {
        throw new Error("Campaign is not active");
      }

      const creatorRepo = new CreatorRepository(client);
      const creator = await creatorRepo.findById(creatorId);
      if (!creator) {
        throw new Error(`Creator ${creatorId} not found`);
      }

      const submissionRepo = new SubmissionRepository(client);
      const count = await submissionRepo.countByCampaignId(campaignId);
      if (count >= campaign.maxSubmissions) {
        throw new Error("Campaign has reached maximum submissions");
      }

      return submissionRepo.create(campaignId, creatorId, contentUrl);
    });
  }

  async approveSubmission(id: number): Promise<Submission> {
    return withTransaction(this.pool, async (client) => {
      const repo = new SubmissionRepository(client);
      const submission = await repo.findById(id);
      if (!submission) {
        throw new Error(`Submission ${id} not found`);
      }
      if (submission.state !== SubmissionState.PENDING) {
        throw new Error("Submission must be in PENDING state to approve");
      }
      return repo.updateState(id, SubmissionState.APPROVED);
    });
  }

  async rejectSubmission(id: number): Promise<Submission> {
    return withTransaction(this.pool, async (client) => {
      const repo = new SubmissionRepository(client);
      const submission = await repo.findById(id);
      if (!submission) {
        throw new Error(`Submission ${id} not found`);
      }
      if (submission.state !== SubmissionState.PENDING) {
        throw new Error("Submission must be in PENDING state to reject");
      }
      return repo.updateState(id, SubmissionState.REJECTED);
    });
  }
}

File: python/src/services/submission_service.py
import asyncpg

from src.data.enums import CampaignState, SubmissionState
from src.data.models import Submission
from src.db.pool import with_transaction
from src.repositories.campaign_repository import CampaignRepository
from src.repositories.creator_repository import CreatorRepository
from src.repositories.submission_repository import SubmissionRepository


class SubmissionService:
    def __init__(self, pool: asyncpg.Pool):
        self.pool = pool

    async def get_submissions_by_campaign(self, campaign_id: int) -> list[Submission]:
        return await SubmissionRepository(self.pool).find_by_campaign_id(campaign_id)

    async def get_submission(self, id: int) -> Submission | None:
        return await SubmissionRepository(self.pool).find_by_id(id)

    async def submit_content(
        self, campaign_id: int, creator_id: str, content_url: str
    ) -> Submission:
        async def txn(conn: asyncpg.Connection) -> Submission:
            campaign = await CampaignRepository(conn).find_by_id(campaign_id)
            if campaign is None:
                raise ValueError(f"Campaign {campaign_id} not found")
            if campaign.state is not CampaignState.ACTIVE:
                raise ValueError("Campaign is not active")

            creator = await CreatorRepository(conn).find_by_id(creator_id)
            if creator is None:
                raise ValueError(f"Creator {creator_id} not found")

            submission_repo = SubmissionRepository(conn)
            count = await submission_repo.count_by_campaign_id(campaign_id)
            if count >= campaign.max_submissions:
                raise ValueError("Campaign has reached maximum submissions")

            return await submission_repo.create(campaign_id, creator_id, content_url)

        return await with_transaction(self.pool, txn)

    async def approve_submission(self, id: int) -> Submission:
        return await self._review(id, SubmissionState.APPROVED, "approve")

    async def reject_submission(self, id: int) -> Submission:
        return await self._review(id, SubmissionState.REJECTED, "reject")

    async def _review(self, id: int, target: SubmissionState, verb: str) -> Submission:
        async def txn(conn: asyncpg.Connection) -> Submission:
            repo = SubmissionRepository(conn)
            submission = await repo.find_by_id(id)
            if submission is None:
                raise ValueError(f"Submission {id} not found")
            if submission.state is not SubmissionState.PENDING:
                raise ValueError(f"Submission must be in PENDING state to {verb}")
            return await repo.update_state(id, target)

        return await with_transaction(self.pool, txn)
